File: systems-for-design-and-implementation/labs/Java_ChatSockets/ChatClient/src/chat/client/gui/ChatClientCtrl.java
package chat.client.gui;

import chat.model.Message;
import chat.model.User;
import chat.services.ChatException;
import chat.services.IChatClient;
import chat.services.IChatServer;

import javax.swing.*;

/**
 * Created by IntelliJ IDEA.
 * User: grigo
 * Date: Mar 18, 2009
 * Time: 2:54:28 PM
 */
public class ChatClientCtrl implements IChatClient
{

    private FriendsListModel  friendsModel;
    private MessagesListModel messagesModel;
    private IChatServer       server;
    private User              user;

    public ChatClientCtrl(IChatServer server)
    {
        this.server = server;
        friendsModel = new FriendsListModel();
        messagesModel = new MessagesListModel();
    }

    public ListModel getFriendsModel()
    {
        return friendsModel;
    }

    public ListModel getMessagesModel()
    {
        return messagesModel;
    }

    public void messageReceived(Message message) throws ChatException
    {
        messagesModel.newMessage(message.getSender().getId(), message.getText());
    }

    public void friendLoggedIn(User friend) throws ChatException
    {
        friendsModel.friendLoggedIn(friend.getId());
    }

    public void friendLoggedOut(User friend) throws ChatException
    {
        friendsModel.friendLoggedOut(friend.getId());
    }

    public void logout()
    {
        try
        {
            server.logout(user, this);
        }
        catch (ChatException e)
        {
            System.out.println("Logout error " + e);
        }
    }

    public void login(String id, String pass) throws ChatException
    {
        User userL = new User(id, pass, "");
        server.login(userL, this);
        user = userL;
        User[] loggedInFriends = server.getLoggedFriends(user);
        System.out.println("Logged friends " + loggedInFriends.length);
        for (User us : loggedInFriends)
        {
            friendsModel.friendLoggedIn(us.getId());
        }

    }

    public void sendMessage(int indexFriend, String txtMsg) throws ChatException
    {
        messagesModel.newMessage(user.getId(), txtMsg);
        User sender = new User(user.getId());
        User receiver = new User((String) friendsModel.getElementAt(indexFriend));
        Message message = new Message(sender, txtMsg, receiver);
        server.sendMessage(message);
    }
}

File: systems-for-design-and-implementation/labs/CSharp_ChatSockets/ChatClient/ChatClientCtrl.cs
using System;
using System.Collections.Generic;
using chat.model;
using chat.services;

namespace chat.client{
    

    public class ChatClientCtrl: IChatObserver
    {
        public event EventHandler<ChatUserEventArgs> updateEvent; //ctrl calls it when it has received an update
        private readonly IChatServer server;
        private User currentUser;
        public ChatClientCtrl(IChatServer server)
        {
            this.server = server;
            currentUser = null;
        }

        public void login(String userId, String pass)
        {
            User user=new User(userId,pass);
            server.login(user,this);
            Console.WriteLine("Login succeeded ....");
            currentUser = user;
            Console.WriteLine("Current user {0}", user);
        }
        public void messageReceived(Message message)
        {
            String mess = "[" + message.Sender.Id + "]: " + message.Text;
            ChatUserEventArgs userArgs=new ChatUserEventArgs(ChatUserEvent.NewMessage,mess);
            Console.WriteLine("Message received");
            OnUserEvent(userArgs);
        }

        public void friendLoggedIn(User friend)
        {
            Console.WriteLine("Friend logged in "+friend);
            ChatUserEventArgs userArgs = new ChatUserEventArgs(ChatUserEvent.FriendLoggedIn, friend.Id);
            OnUserEvent(userArgs);
        }

        public void friendLoggedOut(User friend)
        {
            Console.WriteLine("Friend logged out"+friend);
            ChatUserEventArgs userArgs = new ChatUserEventArgs(ChatUserEvent.FriendLoggedOut, friend.Id);
            OnUserEvent(userArgs);
        }

        public void logout()
        {
            Console.WriteLine("Ctrl logout");
            server.logout(currentUser, this);
            currentUser = null;
        }

        protected virtual void OnUserEvent(ChatUserEventArgs e)
        {
            if (updateEvent == null) return;
            updateEvent(this, e);
            Console.WriteLine("Update Event called");
        }
        public IList<String> getLoggedFriends()
        {
            IList<String> loggedFriends=new List<string>();
            User[] friends = server.getLoggedFriends(currentUser);
            foreach (var user in friends)
            {
                loggedFriends.Add(user.Id);
            }
            return loggedFriends;
        }

        public void sendMessage(string id, string txt)
        {
            //display the sent message on the user window
            String mess = "[" + currentUser.Id + "-->"+id+"]: " + txt;
            ChatUserEventArgs userArgs = new ChatUserEventArgs(ChatUserEvent.NewMessage, mess);
            OnUserEvent(userArgs);
            //sends the message to the server
            User receiver=new User(id);
            Message message=new Message(currentUser,receiver, txt);
            server.sendMessage(message);
        }
    }
}
